src/rt/util/container/hashtab.d

```diff
@@ -328,3 +328,183 @@ unittest
     assert(thrown);
     assert(tab[3] == 3);
 }
+
+/// An open-addressing hash table with quadratic probing
+/// roughly follows the description of Google's dense_hash_map.
+/// WARNING: supports only POD types and no removal by key for now.
+struct FlatHashTab(K, V, alias nullValue, alias hashFunc=hashOf)
+{
+nothrow:
+    this(size_t capacity)
+    {
+        import core.bitop : bsr;
+        if(!capacity)
+            capacity = 32;
+        else
+            capacity = 1<<bsr(capacity);
+        table.length = capacity;
+    }
+
+    @disable this(this);
+
+    ~this()
+    {
+        if(table.length) common.free(table.ptr);
+    }
+
+    void reset()
+    {
+        table[] = Entry.init;
+        items = 0;
+    }
+
+    @property size_t length() const
+    {
+        return items;
+    }
+
+    @property bool empty() const
+    {
+        return items == 0;
+    }
+
+    void opIndexAssign(V value, K key)
+    in
+    {
+        assert(value != nullValue);
+    }
+    body
+    {
+        if(items * 2 == table.length) rehash();
+        size_t h = hashFunc(key);
+        size_t mask = (table.length-1);
+        size_t i = h & mask;
+        if (table[i].value == nullValue)
+        {
+            table[i] = Entry(key, value);
+            items++;
+            return;
+        }
+        if (table[i].key == key)
+        {
+            table[i].value = value;
+            return;
+        }
+        for (size_t n = 1; ; n++)
```

Comment:
[Here](https://github.com/dlang/druntime/blob/68cce532a2ce7954528f801eebd23cc51b410efd/src/rt/aaA.d#L104) is how you loop over triangular numbers.
```d
for (size_t i = h & mask, step = 1;; i = (i + step++) & mask)
{
}
```
No stop condition needed as it will visit every slot.